### PR TITLE
Make ModelBindingContext.Result non-nullable

### PR DESCRIPTION
**Pull request by rynowak**
This change simplifies a bunch of code and fits more in line with the
current design of model binding.

Now, a model binder only has to do anything if it was successful.
'return' is enough to indicate failure.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs [hunk 1]
```diff
@@ -189,13 +189,13 @@ public class ControllerArgumentBinder : IControllerArgumentBinder
             await binder.BindModelAsync(modelBindingContext);
 
             var modelBindingResult = modelBindingContext.Result;
-            if (modelBindingResult != null && modelBindingResult.Value.IsModelSet)
+            if (modelBindingResult != null && modelBindingResult.IsModelSet)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs]: Remove `modelBindingResult != null &&`.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs [hunk 2]
```diff
@@ -85,9 +85,9 @@ public class ControllerArgumentBinder : IControllerArgumentBinder
                 var parameter = parameters[i];
 
                 var result = await BindModelAsync(parameter, controllerContext, valueProvider);
-                if (result != null && result.Value.IsModelSet)
+                if (result != null && result.IsModelSet)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs]: Remove `result != null &&` a few places in this class.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Internal/ModelBindingHelper.cs [hunk 3]
```diff
@@ -450,13 +450,13 @@ public static class ModelBindingHelper
 
             await binder.BindModelAsync(modelBindingContext);
             var modelBindingResult = modelBindingContext.Result;
-            if (modelBindingResult != null && modelBindingResult.Value.IsModelSet)
+            if (modelBindingResult != null && modelBindingResult.IsModelSet)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Internal/ModelBindingHelper.cs]: Remove `modelBindingResult != null &&`.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/BodyModelBinderTests.cs [hunk 4]
```diff
@@ -40,13 +40,13 @@ public class BodyModelBinderTests
...
             // Act
-            var binderResult = await binder.BindModelResultAsync(bindingContext);
+            await binder.BindModelAsync(bindingContext);
 
             // Assert
             mockInputFormatter.Verify(v => v.CanRead(It.IsAny<InputFormatterContext>()), Times.Once);
             mockInputFormatter.Verify(v => v.ReadAsync(It.IsAny<InputFormatterContext>()), Times.Once);
-            Assert.NotNull(binderResult);
-            Assert.True(binderResult.IsModelSet);
+            Assert.NotNull(bindingContext.Result);
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/BodyModelBinderTests.cs]: `null` isn't possible.

Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/BodyModelBinderTests.cs]: Hmm, there's a fair number of these assertions in the new code.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/ArrayModelBinderTest.cs [hunk 5]
```diff
@@ -151,7 +151,7 @@ private static IModelBinder CreateIntBinder()
                     var model = value.ConvertTo(mbc.ModelType);
                     return ModelBindingResult.Success(model);
                 }
-                return null;
+                return ModelBindingResult.Failed();
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/ArrayModelBinderTest.cs]: Why not `default(ModelBindingResult)` everywhere?

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/ArrayModelBinderTest.cs [hunk 6]
```diff
@@ -44,23 +44,23 @@ public class ArrayModelBinderTest
             // Arrange
             var binder = new ArrayModelBinder<string>(new SimpleTypeModelBinder(typeof(string)));
 
-            var context = CreateContext();
-            context.IsTopLevelObject = true;
+            var bindingContext = CreateContext();
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/ArrayModelBinderTest.cs]: Why rename these `context` variables, especially in an already-large PR?

**Comment by kichalla** [comment]
:shipit: after addressing @dougbu's comments

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs]: :+1: - will do a find all references for `.Result` in the product and double check that I didn't miss any.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs]: :+1: - will do a find all references for `.Result` in the product and double check that I didn't miss any.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Internal/ModelBindingHelper.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Internal/ModelBindingHelper.cs]: :+1: - will do a find all references for `.Result` in the product and double check that I didn't miss any.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/BodyModelBinderTests.cs (hunk 4 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/BodyModelBinderTests.cs]: I'll do a global search for these.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/ArrayModelBinderTest.cs (hunk 5 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/ArrayModelBinderTest.cs]: I think this is a little semantically clearer. In general this code is a holdover from when you had to return an `MBR`. In practice neither product nor user code really needs to call `MBR.Failed()`.

Let's start a side discussion about removing `MBR.Failed`.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/ArrayModelBinderTest.cs (hunk 6 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/ArrayModelBinderTest.cs]: It was easier to just rename + find and replace. Tests often used different naming within the same file and this makes it a little bit better.

**Comment by rynowak** [comment]
Updating this as an FYI - I removed a lot of extraneous null checks. if there's any more important feedback I'll address in a follow up

**Pull request closed without merging**